packages/plugin-ext/src/plugin/languages/hover.ts

```diff
@@ -39,20 +50,48 @@ export class HoverAdapter {
         const doc = document.document;
         const pos = Converter.toPosition(position);
 
-        return Promise.resolve(this.provider.provideHover(doc, pos, token)).then(value => {
-            // eslint-disable-next-line @typescript-eslint/no-explicit-any
-            if (!value || !Array.isArray(value.contents) || (value.contents as Array<any>).length === 0) {
-                return undefined;
-            }
-            if (!value.range) {
-                value.range = doc.getWordRangeAtPosition(pos);
-            }
-            if (!value.range) {
-                value.range = new Range(pos, pos);
+        let value: theia.Hover | null | undefined;
+        if (context && context.verbosityRequest) {
+            const previousHoverId = context.verbosityRequest.previousHover.id;
+            const previousHover = this._hoverMap.get(previousHoverId);
+            if (!previousHover) {
+                throw new Error(`Hover with id ${previousHoverId} not found`);
             }
+            const hoverContext: theia.HoverContext = { verbosityDelta: context.verbosityRequest.verbosityDelta, previousHover };
+            value = await this.provider.provideHover(doc, pos, token, hoverContext);
+        } else {
+            value = await this.provider.provideHover(doc, pos, token);
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        if (!value || !Array.isArray(value.contents) || (value.contents as Array<any>).length === 0) {
+            return undefined;
+        }
+        if (!value.range) {
+            value.range = doc.getWordRangeAtPosition(pos);
+        }
+        if (!value.range) {
+            value.range = new Range(pos, pos);
+        }
+
+        const convertedHover: Hover = Converter.fromHover(value);
+        const id = this._hoverCounter;
+        // Check if hover map has more than 10 elements and if yes, remove oldest from the map
+        if (this._hoverMap.size === HoverAdapter.HOVER_MAP_MAX_SIZE) {
+            const minimumId = Math.min(...this._hoverMap.keys());
+            this._hoverMap.delete(minimumId);
+        }
+        this._hoverMap.set(id, value);
+        this._hoverCounter += 1;
```

Comment:
Probably better if the use of the ID and the incrementation of the counter are closer together so that no future code can insert control flow between them and create tricky conditional bugs.